database/db.py

```diff
@@ -115,50 +71,186 @@ def __init__(self, username: str):
+        ...
+
+    @abstractmethod
+    def commit(self) -> None:
+        """Commits the current transaction."""
+        ...
+
+    @abstractmethod
+    def fetchall(self) -> List[tuple]:
+        """Gets all rows of a query result, stores them in the context field `fetched` and also returned."""
+        ...
+
+
+class DBContextManager(ABC):
+
+    @abstractmethod
+    def __enter__(self) -> DBContext: ...
+
+    @abstractmethod
+    def __exit__(self, exc_type, exc_val, exc_tb) -> bool: ...
+
+
+class QueryBuilder(ABC):
+
+    @abstractmethod
+    def get_password_for(self, username) -> str: ...
+
+    @abstractmethod
+    def change_password(self, username, password) -> str: ...
+
+    @abstractmethod
+    def validate_user_exists(self, username) -> str: ...
+
+    @abstractmethod
+    def check_if_friend_request_exists(self, user_id, friends_id) -> str: ...
+
+    @abstractmethod
+    def validate_username_available(self, username) -> str: ...
+
+    @abstractmethod
+    def get_last_user_id(self) -> str: ...
+
+    @abstractmethod
+    def create_user(self, user_id, parsed_data) -> str: ...
+
+    @abstractmethod
+    def create_user_stats(self, user_id) -> str: ...
+
+    @abstractmethod
+    def get_friends_list(self, user_id) -> str: ...
+
+    @abstractmethod
+    def get_user_id(self, username) -> str: ...
+
+    @abstractmethod
+    def get_user_from_id(self, user_id) -> str: ...
+
+    @abstractmethod
+    def get_user_stats(self, user_id) -> str: ...
+
+    @abstractmethod
+    def signin(self, username, token, token_expiration) -> str: ...
+
+    @abstractmethod
+    def get_token(self, username) -> str: ...
+
+    @abstractmethod
+    def get_token_creation_time(self, username) -> str: ...
+
+    @abstractmethod
+    def send_friend_request(self, user_id, friend_id) -> str: ...
+
+    @abstractmethod
+    def add_friend(self, user_id, friend_id) -> str: ...
+
+    @abstractmethod
+    def accept_friend_request(self, user_id, friend_id, accepted_request) -> str: ...
+
+    @abstractmethod
+    def remove_friend(self, user_id, friend_id) -> str: ...
+
+    @abstractmethod
+    def check_for_friend_requests(self, user_id) -> str: ...
+
+    @abstractmethod
+    def revoke_friend_request(self, user_id, friend_id) -> str: ...
+
+    @abstractmethod
+    def logout(self, username) -> str: ...
+
+    @abstractmethod
+    def get_most_games_won(self, number_of_games) -> str: ...
+
+    @abstractmethod
+    def get_longest_win_streak(self, number_of_games) -> str: ...
+
+    @abstractmethod
+    def get_account_info(self, username) -> str: ...
+
+    @abstractmethod
+    def save_account_info(self, username, data) -> str: ...
+
+    @abstractmethod
+    def save_account_info_by_key(self, username, key, value) -> Optional[str]: ...
+
+    @abstractmethod
+    def get_column(self, key) -> str: ...
+
+    @classmethod
+    @abstractmethod
+    def db_check(cls) -> str: ...
+
+
 class DB:
```

Comment:
I stripped out the MySQL implementation from DB and created the MySQLDB class in the mysql_db module which extends DB. To extend DB you just need to create a query builder and a contact manager factory for the new implementation.